[CI][Linter] Replace `flake8` and `isort` with `ruff`

MortalHappiness opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

This PR replaces `flake8` and `isort` with `ruff`.

To minimize changes, all rules that cause file modifications are ignored.

Although this PR involves many files, the changes to Python files are limited to modifications of flake8 comments.

The latest version of `ruff` does not support disabling checks for `SyntaxError`, so `python/ray/serve/tests/test_config_files/syntax_error.py` is explicitly excluded in `pyproject.toml`. See https://docs.astral.sh/ruff/rules/syntax-error/ for details.

The context of why to replace `flake8` with `ruff` is that while running the pre-commit hook of flake8, an error occurs if the Python version is 3.12. This is because the version of flake8 is too old. Then, @jjyao suggested that I replace it with ruff instead of upgrading flake8. See https://github.com/ray-project/ray/issues/47991#issuecomment-2448618658 for details.

Switching to ruff has several benefits:
1. It's faster.
2. Serves as a single tool to replace flake8, isort, and black, making maintenance easier.
3. It offers rich linter rules that can catch more style errors.

Follow-up:
- Enable the ignored rules one by one.
- https://github.com/ray-project/ray/issues/49436
- Run all pre-commit checks.
- Support Cython

## Related issue number

<!-- For example: "Closes #1234" -->
Closes #48508

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

MortalHappiness committed:
[CI][Linter] Migrate to ruff

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Fix SyntaxError files

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Run ruff pre-commit rule in CI

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Update noqa comments to use ruff

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Update requirements_compiled.txt

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Ignore autofixable rules

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

dayshah reviewed (approved):
blazing fast linting ⚡️

MortalHappiness commented on doc/source/train/examples/intel_gaudi/sd.ipynb:
```diff
@@ -342,7 +342,7 @@
     "        assert len(prompt), \"prompt parameter cannot be empty\"\n",
     "\n",
     "        image = self.pipe(prompt, num_inference_steps=50, guidance_scale=7.5).images[0]\n",
-    "\t\treturn image\n",
+    "        return image\n",
```
This fixes `IndentationError`.

MortalHappiness replied:
https://github.com/ray-project/ray/pull/49777

rynewang commented:
cool! @aslonnie any comments?

edoakes commented:
Lonnie is OOO for a few weeks.

@can-anyscale any opposition to this?

can-anyscale commented on ci/lint/lint.sh:
```diff
@@ -10,6 +10,14 @@ clang_format() {
+  # Run pre-commit on all files
+  # TODO(MortalHappiness): Run all pre-commit checks
+  # Currently, we only run the ruff format check
+  pip install -c python/requirements_compiled.txt pre-commit
```
`pre-commit` is not a part of `python/requirements_compiled.txt` I think, you might want to pin a version

MortalHappiness replied:
I found that the Python version in Buildkite is 3.8, while the latest version of `pre-commit` requires Python 3.9 or higher. So, I temporarily pinned `pre-commit` to version 3.5.0. Since Ray no longer supports Python 3.8, perhaps the CI team should consider upgrading the Python version in Buildkite to 3.9 or higher. Once that's done, I'll pin `pre-commit` to the latest version.

can-anyscale commented:
no opposition, but have you tried to run lint locally on linux, macos, etc; does it work?

also can you put more details in the description on why we are replacing things with `ruff`?

MortalHappiness commented:
> no opposition, but have you tried to run lint locally on linux, macos, etc; does it work?

Ran locally on Linux and it works.

> also can you put more details in the description on why we are replacing things with `ruff`?

Sure! Updated!

MortalHappiness committed:
Pin pre-commit version

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

can-anyscale reviewed (approved):
![](https://media3.giphy.com/media/UBBJYj7kgrRpOOdwrv/giphy.gif?cid=5a38a5a26sklkzla83ff25nqwug9dwgj8erws5y8ff26q2ua&ep=v1_gifs_search&rid=giphy.gif&ct=g)

rynewang commented:
Needs Ray Data, RLLib approval, @sven1977 @bveeramani

MortalHappiness committed:
Fix conflicts

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

jjyao commented on ci/lint/format.sh:
```diff
@@ -266,26 +238,11 @@ format_files() {
       black "${BLACK_EXCLUDES[@]}"
     echo "$(date)" "MYPY...."
     mypy_on_each "${MYPY_FILES[@]}"
-    if [ $HAS_FLAKE8 ]; then
-      echo "$(date)" "Flake8...."
-      git ls-files -- '*.py' "${GIT_LS_EXCLUDES[@]}" | xargs -P 5 \
-        flake8 --config=.flake8
-
-      git ls-files -- '*.pyx' '*.pxd' '*.pxi' "${GIT_LS_EXCLUDES[@]}" | xargs -P 5 \
-        flake8 --config=.flake8 "$FLAKE8_PYX_IGNORES"
-    fi
```
Don't we need to replace this with `ruff`?

MortalHappiness replied:
See my comment here https://github.com/ray-project/ray/pull/49435#discussion_r1897546678. These comments were folded.

jjyao replied:
What about `*.py`?

MortalHappiness replied:
I modified the Ruff pre-commit hook to automatically locate the corresponding Python files. See https://github.com/ray-project/ray/pull/49435/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9R34-R36 and https://github.com/astral-sh/ruff-pre-commit/blob/6299d2ca232741e7e33e7fd8a5a10059bd743da7/.pre-commit-hooks.yaml#L6 for details.

jjyao replied:
Does CI run pre-commit hook or it only runs when people enabled pre-commit hook locally?

MortalHappiness replied:
CI runs pre-commit hook here:

- https://buildkite.com/ray-project/premerge/builds/32538#01940ab3-1b33-4d80-ba3c-a62c82ed7f88
- https://github.com/ray-project/ray/pull/49435/files#diff-ad6658a2c0375932b4a9f18312962089115b0d5a0c817e84e0213cc78cbfe927R17-R18

jjyao commented:
Let's wait until getting the approval from @aslonnie

MortalHappiness committed:
Fix conflicts

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on pyproject.toml:
```diff
@@ -0,0 +1,56 @@
+    "C417",
+    "C418",
+    "C419",
+    # Autofixable rules
```
Yes. I'll update the comment.

MortalHappiness commented on doc/source/data/doc_code/preprocessors.py:
```diff
@@ -1,4 +1,4 @@
-# flake8: noqa
+# ruff: noqa
```
Sure. I'll exclude all files in `doc_code` folder.

MortalHappiness committed:
Address comments

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Revert the changes to doc/sources folder

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Exclude doc/source/ folder in pre-commit

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Update comment

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness committed:
Reorganize rules in pyproject.toml

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>

MortalHappiness commented on pyproject.toml:
```diff
@@ -0,0 +1,56 @@
+
+[tool.ruff.lint]
+extend-select = ["I", "B", "Q", "C4" ]
+ignore = [
```
done

MortalHappiness commented:
@aslonnie I have addressed all your comments except for the one about Cython. As mentioned in the description, the E999 rule, which checks for `SyntaxError`, cannot be ignored and is always enabled. Since Cython is a superset of Python, it will always trigger a `SyntaxError`.

MortalHappiness committed:
Remove B020 rule

Closes: ray-project/ray#48508
Signed-off-by: Chi-Sheng Liu <chishengliu@chishengliu.com>
